tensorcircuit/quantum.py

```diff
@@ -1219,6 +1464,193 @@ def quimb2qop(qb_mpo: Any) -> QuOperator:
+
+    :param tn_mpo: MPO in the form of TensorNetwork package
+    :type tn_mpo: ``tn.matrixproductstates.mpo.*``
+    :return: MPO in the form of QuOperator
+    :rtype: QuOperator
+    """
+    tn_tensors = tn_obj.tensors
+    nwires = len(tn_tensors)
+
+    if nwires == 0:
+        return quantum_constructor([], [], [])
+
+    is_mps = all(len(t.shape) <= 3 for t in tn_tensors)
+
+    nodes = []
+    for i in range(nwires):
+        nodes.append(Node(tn_tensors[i], name=f"tensor_{i}"))
+
+    if is_mps:
+        for i in range(nwires - 1):
+            right_edge_idx = len(nodes[i].edges) - 1
+            connect(nodes[i][right_edge_idx], nodes[i + 1][0])
+
+        out_edges = []
+        for i, node in enumerate(nodes):
+            if len(node.edges) == 2:
+                physical_edge = next(e for e in node.edges if e.is_dangling())
+                out_edges.append(physical_edge)
+            else:
+                out_edges.append(node[1])
+
+        in_edges = []
+
+        ignore_edges = []
+        left_dangling = next(
+            (e for e in nodes[0].edges if e.is_dangling() and e not in out_edges), None
+        )
+        right_dangling = next(
+            (e for e in nodes[-1].edges if e.is_dangling() and e not in out_edges), None
+        )
+
+        if left_dangling:
+            ignore_edges.append(left_dangling)
+        if right_dangling:
+            ignore_edges.append(right_dangling)
+
+    else:
+        for i in range(nwires - 1):
+            connect(nodes[i][1], nodes[i + 1][0])
+
+        out_edges = [nodes[i][-1] for i in range(nwires)]
+        in_edges = [nodes[i][-2] for i in range(nwires)]
+        ignore_edges = [nodes[0][0], nodes[-1][1]]
+
+    qop = quantum_constructor(
+        out_edges,
+        in_edges,
+        [],
+        ignore_edges,
+    )
+    return qop
+
+
+def qop2tn(qop: QuOperator) -> Any:
+    """
+    Convert QuOperator back to MPO or MPS in TensorNetwork package.
+
+    :param qop: MPO or MPS in the form of QuOperator, param in docstring
+    :return: MPO or MPS in the form of TensorNetwork
+    :rtype: Union[tn.matrixproductstates.MPO, tn.matrixproductstates.MPS]
+    """
+    sorted_nodes, is_mps, nwires = extract_tensors_from_qop(qop)
+
+    tensors = []
+    for i, node in enumerate(sorted_nodes):
+        edge_to_axis = {edge: axis for axis, edge in enumerate(node.edges)}
+
+        left_v_ax, right_v_ax = None, None
+
+        if i > 0:
+            prev_node = sorted_nodes[i - 1]
+            edge_to_prev = next(
+                e for e in node.edges if (e.node1 is prev_node or e.node2 is prev_node)
+            )
+            left_v_ax = edge_to_axis[edge_to_prev]
+
+        if i < nwires - 1:
+            next_node = sorted_nodes[i + 1]
+            edge_to_next = next(
+                e for e in node.edges if (e.node1 is next_node or e.node2 is next_node)
+            )
+            right_v_ax = edge_to_axis[edge_to_next]
```

Comment:
are these checks necessary? as the sorted_node has already sorted with predictable edge order?